### PR TITLE
[Bridge handler] Added a reserved period for bridge tx resending

### PR DESCRIPTION
**Pull request by hyunsooda**
## Proposed changes

A value transfer direction from child to parent has a loop that repeatedly requests bridge transactions (value transfer and anchoring transactions) if the corresponding receipt does not arrive. Its duplicated transaction leads to pay unnecessary operator's gas costs. This PR added a reserved period (30s), where the resending should be triggered. Currently, the value transfer recovery implementation that has the wrong semantic can send one duplicated transaction. This duplicated sending is not associated to the bridge v2 implementation (#1565)

Issue:
- https://github.com/klaytn/klaytn/pull/1295#issuecomment-1177082342
- #1579 
- https://forum.klaytn.foundation/t/baobob-invalid-message/5813?u=jack_jin
## Types of changes

Please put an x in the boxes related to your change.

- [x] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: node/sc/sub_bridge_handler.go [hunk 1]
```diff
@@ -383,26 +390,18 @@ func (sbh *SubBridgeHandler) handleParentChainInvalidTxResponseMsg(msg p2p.Msg)
 	if err := msg.Decode(&invalidTxs); err != nil && err != rlp.EOL {
 		return errResp(ErrDecode, "msg %v: %v", msg, err)
 	}
-	txPool := sbh.subbridge.GetBridgeTxPool()
 	for _, invalidTx := range invalidTxs {
-		if tx := txPool.Get(invalidTx.TxHash); tx != nil {
-			logger.Error("A bridge tx was not executed", "err", invalidTx.ErrStr,
+		if tx := sbh.subbridge.GetBridgeTxPool().Get(invalidTx.TxHash); tx != nil {
```
Comment (hqjang-pepper) [inline node/sc/sub_bridge_handler.go:394]:
```suggestion
		if tx := txPool.Get(invalidTx.TxHash); tx != nil {
```

File: node/sc/sub_bridge_handler.go [hunk 2]
```diff
@@ -383,26 +390,18 @@ func (sbh *SubBridgeHandler) handleParentChainInvalidTxResponseMsg(msg p2p.Msg)
 	if err := msg.Decode(&invalidTxs); err != nil && err != rlp.EOL {
 		return errResp(ErrDecode, "msg %v: %v", msg, err)
 	}
-	txPool := sbh.subbridge.GetBridgeTxPool()
 	for _, invalidTx := range invalidTxs {
```
Comment (hqjang-pepper) [inline node/sc/sub_bridge_handler.go:393]:
```suggestion
        txPool := sbh.subbridge.GetBridgeTxPool()
	for _, invalidTx := range invalidTxs {
```
I think we don't have to get txPool newly when we loop everytime.

File: node/sc/sub_bridge_handler.go (hunk 1 above)
Comment (hyunsooda) [reply node/sc/sub_bridge_handler.go]: Thanks. changed.

File: node/sc/sub_bridge_handler.go (hunk 2 above)
Comment (hyunsooda) [reply node/sc/sub_bridge_handler.go:394]: Thanks. changed.

**Comment by hyunsooda** [comment]
Fixed a logical error. Thank you, @2dvorak. PTAL again.

File: node/sc/sub_bridge_handler.go [hunk 3]
```diff
@@ -600,3 +614,16 @@ func (sbh *SubBridgeHandler) WriteReceiptFromParentChain(blockHash common.Hash,
 func (sbh *SubBridgeHandler) GetReceiptFromParentChain(blockHash common.Hash) *types.Receipt {
 	return sbh.subbridge.chainDB.ReadReceiptFromParentChain(blockHash)
 }
+
+func (sbh *SubBridgeHandler) removeBridgeTx(tx *types.Transaction) {
+	if err := sbh.subbridge.GetBridgeTxPool().RemoveTx(tx); err != nil {
+		logger.Error("[SC][HandleResponse] Failed to remove bridge tx",
+			"txType", tx.Type(), "txNonce", tx.Nonce(), "txHash", tx.Hash().String())
```
[the remark this answers is not in the document]
Comment (hyunsooda) [reply node/sc/sub_bridge_handler.go:623]: Resolved via another channel

**Review by 2dvorak — COMMENTED** [review]
Tested with the following steps:
1. Deploy parent and child chain
2. Deploy bridge & token contract and subscribe event
3. Stop CN
4. Request value transfer from child -> parent

Before this PR, I observed several "known transaction" error messages because the subbridge constantly resends the same tx until it gets the corresponding receipt. Due to inter-chain latency, those errors were produced even when I did not stopped CN in step 3.

After this PR, I could not see any "known transaction" error being produced, and even when I stopped the CN in step 3, I could observe only 1 "known transaction" error. That unhandled tx was handled in the next period of VT recovery (if enabled) ~~or next value transfer event.~~

Just one question - without VT recovery, if the CN is down, the value transfer event will be retried only once after resend interval. ~~If the CN is still down, that unhandled event will be retried when another value transfer request is made.~~ Is this intended?

**EDIT** Striked out some wrong comments

**Comment by hyunsooda** [comment]
@2dvorak, Let me explain why you observed them. First, the only one-time resending is because of the `known transaction` by txpool policy. The `EN` already holds a transaction and if a duplicated transaction comes, that duplicated transaction is dropped (You can see from the `EN`s log) and sends an invalid tx message to the bridge node. The bridge node receives the invalid message and removes that transaction from bridge txpool. Thus, no resending is created from now on.

Could you recheck this one? `If the CN is still down, that unhandled event will be retried when another value transfer request is made`. The only new bridge transaction is sent if the previously unhandled transaction was removed.

**Comment by 2dvorak** [comment]
@hyunsooda OK, thank you for the explanation.

And I was wrong about `If the CN is still down, that unhandled event will be retried when another value transfer request is made`. I checked again and observed that the VT failed due to CN being down was handled after CN went alive, regardless of new VT request being made. Sorry for confusing you, I will edit my above comment.

**Review by 2dvorak — APPROVED** [review]

**Comment by hyunsooda** [comment]
@2dvorak, I slightly touched the resending logic. As the value transfer recovery (VTR) makes a new tx and sends it in an independent thread, I think we do not need to manage those txs for resending. However, the anchoring txs are not captured by the VTR and thus we should manage only anchoring txs for resending. The new change removes value transfer txs right after they are sent, but not anchoring txs.

**Comment by hyunsooda** [comment]
pending

**Comment by hyunsooda** [comment]
pending

**Pull request closed without merging**